# 🦙 Ollama

<center>
    <img src="https://ollama.com/public/cloud.png" alt="ollama is cool" width="200"/>
    <img src="https://mintcdn.com/ollama-9269c548/w-L7kuDqk3_8zi5c/images/welcome.png?w=840&fit=max&auto=format&n=w-L7kuDqk3_8zi5c&q=85&s=cc3770b69817332b5961d0faecdce473" alt="ollama working" width="400"/>
</center>



As we can see in these picture (^^), [Ollama](https://docs.ollama.com/) is a cool, simple and lightweight tool that allows LLMs to be run locally without complex configuration. 

### 🏳️ Open the cc-in2p3 Jupyter Notebook launcher at 

> `https://notebook.cc.in2p3.fr`


launch a new terminal session

<center>
    <img src="https://icon-library.com/images/bash-icon/bash-icon-6.jpg" alt="bash is also cool" width="200"/>
</center>


and execute therein the following commands to load the previously prepared enviroment with ollama installed

```bash
$ module load ollama
```

launch an ollama server

```bash
$ export OLLAMA_MODELS=/pbs/throng/training/universite-hiver/wschooloptim/cache/ollama/models
$ export CUDA_VISIBLES_DEVICES=GPU-99ec877c-53b2-49ba-a1ea-0b6f0d0f0c08; 
$ export OLLAMA_HOST=http://127.0.0.1:11435; 
$ export OLLAMA_FLASH_ATTENTION=1; 
$ ollama serve &
```

and pull `zephyr` model

```bash
 $ ollama pull zephyr
```

Check that `zephyr` model is present and show some informations

```bash
$ ollama list
$ ollama show zephyr
```

Create one request to ollama's server:

```bash
$ curl -X POST http://localhost:11435/v1/chat/completions \
-H "Content-Type: application/json" \
-d '{
  "model": "zephyr:7b",
  "messages": [{ "role": "user", "content": "Who won the last presidential elections in France?" }]
}'
```

Keep that terminal open for the rest of the tutorial, we will check therein ollama server logs.

### 🏴 Open another cc-in2p3 Jupyter Notebook launcher at 

> `https://notebook.cc.in2p3.fr`


and launch now the Notebook called `Python 3.13 (Ollama)`.

We will work from now in that notebook.

In [1]:
! /pbs/throng/training/universite-hiver/wschooloptim/vllm/bin/xan tail /pbs/home/u/univ-hiver01/private/resinwinterschool/tweets_spanish_english_25K.csv | /pbs/throng/training/universite-hiver/wschooloptim/vllm/bin/xan v 


Displaying 3 cols from 10 rows of <stdin>
┌───┬───────┬───────────────────┬──────────────────────────────────────────┐
│ - │ index │ text              │ english                                  │
├───┼───────┼───────────────────┼──────────────────────────────────────────┤
│ 0 │ 24990 │ RT @joseantoniok… │ RT @joseantoniokast: There is no time t… │
│ 1 │ 24991 │ RT @reloaled: Pe… │ But look, the Minister is walking with … │
│ 2 │ 24992 │ RT @25Hernan25: … │ RT @25Hernan25: @verge_cdmx is the new … │
│ 3 │ 24993 │ RT @Sinfiltros_t… │ RT @Sinfiltros_tv: "Chilenians want a f… │
│ 4 │ 24994 │ RT @hernan_sr: I… │ RT @hernan_sr: It's amazing to see a yo… │
│ 5 │ 24995 │ RT @GAMBA_CL: CA… │ RT @GAMBA_CL: CADEM:Chilians see Kast a… │
│ 6 │ 24996 │ RT @delosquesobr… │ RT @delosquesobran: 🔴Daniel Stingo and… │
│ 7 │ 24997 │ RT @Daniel399492… │ RT @Daniel399492892: Bernardo Fontaine,… │
│ 8 │ 24998 │ RT @joseantoniok… │ RT @joseantoniokast: There's no time to… │
│ 9 │ 24999 │ RT @csantander23… │ 

##### Import python packages

In [2]:
import csv
import time
import asyncio
from openai import OpenAI
from string import Template
import nest_asyncio
nest_asyncio.apply()

##### Load inputs and create asynchronous iterator

In [3]:
input_file = "/pbs/home/u/univ-hiver01/private/resinwinterschool/tweets_spanish_english_25K.csv"
content_column = 'english'

start = time.time()
with open(input_file, 'r') as f:
    tweets = [d[content_column] for d in csv.DictReader(f)]
end = time.time()
print(f"Loading {len(tweets)} tweets tooks {end - start} seconds.")


Loading 25000 tweets tooks 0.1273493766784668 seconds.


##### Load choices and extra body dictionary with structured_outputs options

In [4]:
choices_file = "/pbs/throng/training/universite-hiver/wschooloptim/resinwinterschool/multiple.txt"
with open(choices_file, 'r') as f:
    choice = f.read().split("\n")

extra_body = {"structured_outputs": {"choice": choice}}

print(extra_body)

{'structured_outputs': {'choice': ['Jara', 'Kast', 'Kaiser', 'Matthei', 'Parisi', 'Artés', 'Mayne-Nichols', 'Enríquez-Ominami', 'None']}}


##### Load instructions

In [5]:
instructions_file = "/pbs/throng/training/universite-hiver/wschooloptim/resinwinterschool/instructions.txt"
with open(instructions_file, 'r') as f:
    instructions = f.read()
print(instructions)

Please classify the following social media message (posted in the weeks leading up to the 2025 Chilean presidential election) according to whether it explicitly expresses the intention  of the author to vote for or calls for a vote for any of the candidates in that election: Jeannette Jara, José Antonio Kast, Johannes Kaiser, Evelyn Matthei, Franco Parisi, Eduardo Artés, Harold Mayne-Nichols, Marco Enríquez-Ominami (also known as MEO), or whether it expresses neither of these intentions. Your answer should be based solely on the information contained in the message. Do not confuse a simple mention of a candidate with an intention or call to vote for him. Do not assume that a message containing only positive opinions about a particular candidate explicitly expresses the intention to vote for that candidate. Do not assume that a message corresponding to the opinions or political positions of a particular candidate necessarily expresses the intention to vote for that candidate. Do not con

##### Create ollama openai client

In [6]:
client = OpenAI(
    base_url="http://127.0.0.1:11435/v1/", # this is the same HOST given at ollama server launching
    api_key="EMPTY"  # required but unused
)

##### Create async functions to asynchronously request llm server


In [7]:
async def inputsIterator(start, end):
    for tweet in tweets[start:end]:
        yield tweet

In [8]:
async def doCompletetion(tweet):
    return client.chat.completions.create(
        model=client.models.list().data[0].id,
        messages=[{
            'role': 'user',
            'content': Template(instructions).substitute(tweet=tweet)
        }],
        extra_body=extra_body,
        max_completion_tokens=16,
        temperature=0.7,
        top_p=0.95,
    )

In [9]:
async def run_all(start, end):
    # Asynchronously call the function for each prompt
    tasks = [
        doCompletetion(tweet)
        async for tweet in inputsIterator(start, end)
    ]
    # Gather and run the tasks concurrently
    results = await asyncio.gather(*tasks)
    return results

In [10]:
# Run all courutines
init_idx = 1111
nb_tweets = 500

start = time.time()
results = asyncio.run(run_all(start=init_idx, end=init_idx + nb_tweets))
elapsed = time.time() - start

whole_db_nb_tweets = 1043873


print(f"""
Annotationg {nb_tweets} tweets took {elapsed} seconds,
this is {elapsed / nb_tweets} seconds per tweet or
{(1 / (24 * 3600)) * whole_db_nb_tweets * elapsed / nb_tweets} days for the whole database of {whole_db_nb_tweets} inputs.
""")


Annotationg 500 tweets took 310.1631088256836 seconds,
this is 0.6203262176513672 seconds per tweet or
7.494696641185019 days for the whole database of 1043873 inputs.



> Annotationg 500 tweets took 310.1631088256836 seconds,
>
> this is 0.6203262176513672 seconds per tweet or
>
> 7.494696641185019 days for the whole database of 1043873 inputs.

In [12]:
for tweet, answer in zip(tweets, [r.choices[0].message.content for r in results]):
    if answer == "Kast":
        print("-----------------------------------------------" + "\n" + tweet + "\n" + answer)

-----------------------------------------------
RT @DRESTRUM__Pl: 🔴 Chileans consider Kast to be the best choice for:

- improving economy - creating employment
- combating crime and drug trafficking
- providing governance
- controlling immigration
- ending corruption
- solving country's problems
- facing crises

UN München. https://t.co/ExWLUhkaH3
Kast
-----------------------------------------------
RT @hernan_sr: 🚨ATTENTION

Jose Antonio Kast's proposal would affect more than 1.2 million retirees in Chile with 80 fewer Lucas per month.

Because Jeannette Jara led the pension reform, now more than ever we must defend it.

#IncreasePensions https://t.co/NuGN62kDiB
Kast
-----------------------------------------------
In simple terms, Kast finds it soft to govern and prefers to be in opposition. There's no other way to explain that his proposals on various sensitive issues for the average Chilean, are counter-proposals.
#MesaCentral
#enacional #Cadem
Citizen Panel - Matthei - Kaiser
#Jar